lib/global_uid/base.rb

```diff
@@ -29,169 +20,104 @@ def self.create_uid_tables(id_table_name, options={})
-
-      begin
-        Timeout.timeout(connection_timeout, ConnectionTimeoutException) do
-          ActiveRecord::Base.mysql2_connection(config)
-        end
-      rescue ConnectionTimeoutException => e
-        notify e, "Timed out establishing a connection to #{name}"
-        nil
-      rescue Exception => e
-        notify e, "establishing a connection to #{name}: #{e.message}"
-        nil
+      with_servers do |server|
+        server.connection.execute("DROP TABLE IF EXISTS `#{id_table_name}`")
       end
     end
 
     def self.init_server_info
-      id_servers = self.global_uid_servers
-      increment_by = self.global_uid_options[:increment_by]
+      id_servers = GlobalUid.configuration.id_servers
+      increment_by = GlobalUid.configuration.increment_by
+      connection_timeout = GlobalUid.configuration.connection_timeout
 
       raise "You haven't configured any id servers" if id_servers.nil? or id_servers.empty?
       raise "More servers configured than increment_by: #{id_servers.size} > #{increment_by} -- this will create duplicate IDs." if id_servers.size > increment_by
 
-      id_servers.map do |name, i|
-        info = {}
-        info[:cx]       = nil
-        info[:name]     = name
-        info[:retry_at] = nil
-        info[:rand]     = rand
-        info[:new?]     = true
-        info
+      id_servers.map do |name|
+        GlobalUid::Server.new(name,
+          increment_by: increment_by,
+          connection_timeout: connection_timeout
+        )
       end
     end
 
-    def self.disconnect!
-      self.servers = nil
-    end
-
     def self.setup_connections!
-      connection_timeout = self.global_uid_options[:connection_timeout]
-
-      if self.servers.nil?
-        self.servers = init_server_info
-        # sorting here sets up each process to have affinity to a particular server.
-        self.servers = self.servers.sort_by { |s| s[:rand] }
-      end
-
-      self.servers.each do |info|
-        next if info[:cx]
-
-        if info[:new?] || ( info[:retry_at] && Time.now > info[:retry_at] )
-          info[:new?] = false
-
-          connection = new_connection(info[:name], connection_timeout)
-          info[:cx]  = connection
-          info[:retry_at] = Time.now + self.global_uid_options[:connection_retry] if connection.nil?
-        end
-      end
+      self.servers ||= init_server_info
+      self.servers.each(&:connect)
+    end
 
-      self.servers
+    def self.disconnect!
+      self.servers.each(&:disconnect!) unless servers.nil?
+      self.servers = nil
     end
 
-    def self.with_connections
+    def self.with_servers
       servers = setup_connections!
-
-      if !self.global_uid_options[:per_process_affinity]
-        servers = servers.sort_by { rand } #yes, I know it's not true random.
-      end
+      servers = servers.shuffle if !!GlobalUid.configuration.per_process_affinity?
 
       raise NoServersAvailableException if servers.empty?
```

Comment:
We should also call notify here, so the client is informed. E.g.

`GlobalUid.configuration.notifier.notify(NoServersAvailableException.new("No servers found")`